src/models/pretraining.py

```diff
@@ -0,0 +1,115 @@
+import torch.nn as nn
+import torch
+import numpy as np
+from torch.nn import CTCLoss
+import lpips
+import matplotlib.pyplot as plt
+
+from src.utils.common import PositionalEncoding, compute_ctc_loss, masking
+from src.models.vtr.ocr import OCRHead
+
+
+class Pretrain(nn.Module):
+    def __init__(
+        self,
+        emb_size: int = 512,
+        n_head: int = 8,
+        n_layers: int = 4,
+        device: str = "cpu",
+        ocr: OCRHead = None,
+        char2int_dict: dict = None,
+        alpha: float = 1,
+    ):
+        super().__init__()
+        self.encoder = nn.TransformerEncoder(
+            nn.TransformerEncoderLayer(d_model=emb_size, nhead=n_head, dim_feedforward=emb_size), num_layers=n_layers
+        )
+        self.positional_enc = PositionalEncoding(emb_size)
+        self.positional_dec = PositionalEncoding(emb_size)
+        # self.masking = SpanMaskingGenerator(emb_size)
+        self.linear = nn.Linear(emb_size, emb_size)
+        self.decoder = nn.TransformerEncoder(
+            nn.TransformerEncoderLayer(d_model=emb_size, nhead=n_head, dim_feedforward=emb_size), num_layers=n_layers
+        )
+        self.ocr = ocr
+        self.ctc_criterion = CTCLoss(reduction="sum", zero_infinity=True)
+        self.char2int_dict = char2int_dict
+        self.device = device
+        self.criterion = lpips.LPIPS(net="vgg").to(device)
+        # self.criterion = nn.MSELoss()
+        self.emb_size = emb_size
+        self.alpha = alpha
+        self.iter = 0
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, input_batch: dict[str, list | torch.Tensor]):
+
+        batch_size, slice_count, height, width = input_batch["slices"].shape
+        slices = input_batch["slices"].view(batch_size, slice_count, height * width).clone()
+
+        grey_slice = torch.full((height, width), 128, dtype=torch.float32, device=self.device).flatten()
+        masks = []
+        unmasked_slices = []
+        unmasked_texts = []
+        for i in range(batch_size):
+            not_padded_len = torch.sum(input_batch["attention_mask"][i], dtype=torch.int).item()
+            mask = masking(not_padded_len)
+            masks.append(mask)
+            masked_idx = np.where(mask == 1)[0]
+            slices[i][masked_idx] = grey_slice
```

Comment:
See comments in `common.py` file, I provide effective implementation to create mask for entire batch.
So:

```python
# [batch size; slice count], `True` for patches to mask 
mask = create_noise_mask(batch_size, slice_count, self.noise_density, self.span_lens)
mask *= input_batch["attention_mask"]  # attn mask has the same shape and `True` for non padding values

slices[mask, :, :] = self.MASK_VALUE  # this would fill all masked tokens with specified value, e.g 128 
```